File: src/main/java/me/dyzjct/kura/module/modules/render/PopChams.java
package me.dyzjct.kura.module.modules.render;

import com.mojang.authlib.GameProfile;
import me.dyzjct.kura.event.events.client.PacketEvents;
import me.dyzjct.kura.event.events.render.RenderEvent;
import me.dyzjct.kura.module.Category;
import me.dyzjct.kura.module.Module;
import me.dyzjct.kura.setting.BooleanSetting;
import me.dyzjct.kura.setting.IntegerSetting;
import me.dyzjct.kura.setting.Setting;
import me.dyzjct.kura.utils.gl.MelonTessellator;
import me.dyzjct.kura.utils.render.TotemPopCham;
import net.minecraft.client.entity.EntityOtherPlayerMP;
import net.minecraft.client.model.ModelBase;
import net.minecraft.client.model.ModelPlayer;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.network.play.server.SPacketEntityStatus;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import org.lwjgl.opengl.GL11;

import java.awt.*;

@Module.Info(name="PopChams", description="Renders when someone pops", category= Category.RENDER)
public class PopChams
        extends Module {
    public static PopChams INSTANCE = new PopChams();
    public BooleanSetting self = this.bsetting("Self", false);
    public IntegerSetting rF = this.isetting("RedFill", 255, 0, 255);
    public IntegerSetting gF = this.isetting("GreenFill", 26, 0, 255);
    public IntegerSetting bF = this.isetting("BlueFill", 42, 0, 255);
    public IntegerSetting aF = this.isetting("AlphaFill", 42, 0, 255);
    public IntegerSetting fadestart = this.isetting("FadeStart", 200, 0, 3000);
    public Setting<Double> fadetime = this.dsetting("FadeTime", 0.5, 0.0, 2.0);
    public BooleanSetting onlyOneEsp = this.bsetting("OnlyOneEsp", true);
    EntityOtherPlayerMP player;
    ModelPlayer playerModel;
    Long startTime;
    double alphaFill;

    public static Color newAlpha(Color color, int alpha) {
        return new Color(color.getRed(), color.getGreen(), color.getBlue(), alpha);
    }

    public static void glColor(Color color) {
        GL11.glColor4f((float)((float)color.getRed() / 255.0f), (float)((float)color.getGreen() / 255.0f), (float)((float)color.getBlue() / 255.0f), (float)((float)color.getAlpha() / 255.0f));
    }

    @SubscribeEvent
    public void onPacketReceive(PacketEvents.Receive event) {
        SPacketEntityStatus packet;
        if (event.getPacket() instanceof SPacketEntityStatus && (packet = (SPacketEntityStatus)event.getPacket()).getOpCode() == 35 && packet.getEntity((World)PopChams.mc.world) != null && (((Boolean)this.self.getValue()).booleanValue() || packet.getEntity((World)PopChams.mc.world).getEntityId() != PopChams.mc.player.getEntityId())) {
            GameProfile profile = new GameProfile(PopChams.mc.player.getUniqueID(), "");
            this.player = new EntityOtherPlayerMP((World)PopChams.mc.world, profile);
            this.player.copyLocationAndAnglesFrom(packet.getEntity((World)PopChams.mc.world));
            this.playerModel = new ModelPlayer(0.0f, false);
            this.startTime = System.currentTimeMillis();
            this.playerModel.bipedHead.showModel = false;
            this.playerModel.bipedBody.showModel = false;
            this.playerModel.bipedLeftArmwear.showModel = false;
            this.playerModel.bipedLeftLegwear.showModel = false;
            this.playerModel.bipedRightArmwear.showModel = false;
            this.playerModel.bipedRightLegwear.showModel = false;
            this.alphaFill = ((Integer)this.aF.getValue()).intValue();
        }
    }

    @Override
    public void onWorldRender(RenderEvent event) {
        if (((Boolean)this.onlyOneEsp.getValue()).booleanValue()) {
            if (this.player == null || PopChams.mc.world == null || PopChams.mc.player == null) {
                return;
            }
            GL11.glLineWidth((float)1.0f);
            Color fillColorS = new Color((Integer)this.rF.getValue(), (Integer)this.bF.getValue(), (Integer)this.gF.getValue(), (Integer)this.aF.getValue());
            int fillA = fillColorS.getAlpha();
            long time = System.currentTimeMillis() - this.startTime - ((Number)this.fadestart.getValue()).longValue();
            if (System.currentTimeMillis() - this.startTime > ((Number)this.fadestart.getValue()).longValue()) {
                double normal = this.normalize(time, 0.0, ((Number)this.fadetime.getValue()).doubleValue());
                normal = MathHelper.clamp((double)normal, (double)0.0, (double)1.0);
                normal = -normal + 1.0;
                fillA *= (int)normal;
            }
            Color fillColor = PopChams.newAlpha(fillColorS, fillA);
            if (this.player != null && this.playerModel != null) {
                MelonTessellator.INSTANCE.prepare(7);
                GL11.glPushAttrib((int)1048575);
                GL11.glEnable((int)2881);
                GL11.glEnable((int)2848);
                if (this.alphaFill > 1.0) {
                    this.alphaFill -= this.fadetime.getValue().doubleValue();
                }
                Color fillFinal = new Color(fillColor.getRed(), fillColor.getGreen(), fillColor.getBlue(), (int)this.alphaFill);
                PopChams.glColor(fillFinal);
                GL11.glPolygonMode((int)1032, (int)6914);
                TotemPopCham.renderEntity((EntityLivingBase)this.player, (ModelBase)this.playerModel, this.player.limbSwing, this.player.limbSwingAmount, 1.0f);
                GL11.glPolygonMode((int)1032, (int)6913);
                TotemPopCham.renderEntity((EntityLivingBase)this.player, (ModelBase)this.playerModel, this.player.limbSwing, this.player.limbSwingAmount, 1.0f);
                GL11.glPolygonMode((int)1032, (int)6914);
                GL11.glPopAttrib();
                MelonTessellator.release();
            }
        } else if (!((Boolean)this.onlyOneEsp.getValue()).booleanValue()) {
            new TotemPopCham(this.player, this.playerModel, this.startTime, this.alphaFill);
        }
    }

    public double normalize(double value, double min, double max) {
        return (value - min) / (max - min);
    }
}

File: src/main/java/me/windyteam/kura/module/modules/render/PopChams.kt
package me.windyteam.kura.module.modules.render

import com.mojang.authlib.GameProfile
import me.windyteam.kura.event.events.client.PacketEvents
import me.windyteam.kura.event.events.render.RenderEvent
import me.windyteam.kura.module.Category
import me.windyteam.kura.module.Module
import me.windyteam.kura.setting.Setting
import me.windyteam.kura.utils.gl.MelonTessellator.prepare
import me.windyteam.kura.utils.gl.MelonTessellator.release
import me.windyteam.kura.utils.render.TotemPopCham
import net.minecraft.client.entity.EntityOtherPlayerMP
import net.minecraft.client.model.ModelBase
import net.minecraft.client.model.ModelPlayer
import net.minecraft.entity.EntityLivingBase
import net.minecraft.network.Packet
import net.minecraft.network.play.server.SPacketEntityStatus
import net.minecraft.util.math.MathHelper
import net.minecraft.world.World
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import org.lwjgl.opengl.GL11
import java.awt.Color

@Module.Info(name = "PopChams", description = "Renders when someone pops", category = Category.RENDER)
class PopChams : Module() {
    var self = bsetting("Self", false)
    @JvmField
    var rF = isetting("RedFill", 255, 0, 255)
    @JvmField
    var gF = isetting("GreenFill", 26, 0, 255)
    @JvmField
    var bF = isetting("BlueFill", 42, 0, 255)
    @JvmField
    var aF = isetting("AlphaFill", 42, 0, 255)
    @JvmField
    var fadestart = isetting("FadeStart", 200, 0, 3000)
    @JvmField
    var fadetime: Setting<Double> = dsetting("FadeTime", 0.5, 0.0, 2.0)
    var onlyOneEsp = bsetting("OnlyOneEsp", true)
    var player: EntityOtherPlayerMP? = null
    var playerModel: ModelPlayer? = null
    var startTime: Long? = null
    var alphaFill = 0.0

    @SubscribeEvent
    fun onPacketReceive(event: PacketEvents.Receive) {
        if (event.getPacket<Packet<*>>() is SPacketEntityStatus && event.getPacket<SPacketEntityStatus>()
                .opCode.toInt() == 35
        ) {
            event.getPacket<SPacketEntityStatus>().getEntity(mc.world as World)
            if (self.value as Boolean || event.getPacket<SPacketEntityStatus>().getEntity(mc.world).getEntityId() != mc.player.getEntityId()) {
                val profile = GameProfile(mc.player.uniqueID, "")
                player = EntityOtherPlayerMP(mc.world, profile)
                player!!.copyLocationAndAnglesFrom(event.getPacket<SPacketEntityStatus>().getEntity(mc.world as World))
                playerModel = ModelPlayer(0.0f, false)
                startTime = System.currentTimeMillis()
                playerModel!!.bipedHead.showModel = false
                playerModel!!.bipedBody.showModel = false
                playerModel!!.bipedLeftArmwear.showModel = false
                playerModel!!.bipedLeftLegwear.showModel = false
                playerModel!!.bipedRightArmwear.showModel = false
                playerModel!!.bipedRightLegwear.showModel = false
                alphaFill = aF.value.toDouble()
            }
        }
    }

    override fun onWorldRender(event: RenderEvent) {
        if (onlyOneEsp.value) {
            if (player == null || mc.world == null || mc.player == null) {
                return
            }
            GL11.glLineWidth(1.0f)
            val fillColorS = Color((rF.value as Int), (bF.value as Int), (gF.value as Int), (aF.value as Int))
            var fillA = fillColorS.alpha
            val time = System.currentTimeMillis() - startTime!! - (fadestart.value as Number).toLong()
            if (System.currentTimeMillis() - startTime!! > (fadestart.value as Number).toLong()) {
                var normal = this.normalize(time.toDouble(), 0.0, (fadetime.value as Number).toDouble())
                normal = MathHelper.clamp(normal, 0.0, 1.0)
                normal = -normal + 1.0
                fillA *= normal.toInt()
            }
            val fillColor = newAlpha(fillColorS, fillA)
            if (player != null && playerModel != null) {
                prepare(7)
                GL11.glPushAttrib(1048575)
                GL11.glEnable(2881)
                GL11.glEnable(2848)
                if (alphaFill > 1.0) {
                    alphaFill -= fadetime.value
                }
                val fillFinal = Color(fillColor.red, fillColor.green, fillColor.blue, alphaFill.toInt())
                glColor(fillFinal)
                GL11.glPolygonMode(1032, 6914)
                TotemPopCham.renderEntity(
                    player as EntityLivingBase?,
                    playerModel as ModelBase?,
                    player!!.limbSwing,
                    player!!.limbSwingAmount,
                    1.0f
                )
                GL11.glPolygonMode(1032, 6913)
                TotemPopCham.renderEntity(
                    player as EntityLivingBase?,
                    playerModel as ModelBase?,
                    player!!.limbSwing,
                    player!!.limbSwingAmount,
                    1.0f
                )
                GL11.glPolygonMode(1032, 6914)
                GL11.glPopAttrib()
                release()
            }
        } else if (!(onlyOneEsp.value as Boolean)) {
            TotemPopCham(player, playerModel, startTime, alphaFill)
        }
    }

    fun normalize(value: Double, min: Double, max: Double): Double {
        return (value - min) / (max - min)
    }

    companion object {
        @JvmField
        var INSTANCE = PopChams()
        fun newAlpha(color: Color, alpha: Int): Color {
            return Color(color.red, color.green, color.blue, alpha)
        }

        fun glColor(color: Color) {
            GL11.glColor4f(
                color.red.toFloat() / 255.0f,
                color.green.toFloat() / 255.0f,
                color.blue.toFloat() / 255.0f,
                color.alpha.toFloat() / 255.0f
            )
        }
    }
}